### PR TITLE
ASoC: SOF: pm: Fix prepare callback behavior for OF usecase

### PR DESCRIPTION
**Pull request by dbaluta**
On i.MX platforms PM is not managed via ACPI although CONFIG_ACPI
can be set. So, in order to set correctly set the system target state
we introduce a flag for ACPI capable platforms.

For ACPI capable platforms the behavior of prepare callback remains
unchanged while for non-ACPI capable platforms system target state
is always set to default SOF_SUSPEND_S3.

Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>

### CODE REVIEW

**Review by paulstelian97 — COMMENTED** [review]
Waiting for the i.MX ACPI-capable=false change.

**Comment by dbaluta** [comment]
@paulstelian97 I preferred to make as less changes as possible relying on the fact that by default global structures are zeroed. So, acpi_capable will be default set to false if missing in the initialization.

Do you think we should explictly set it to false?

**Comment by paulstelian97** [comment]
> @paulstelian97 I preferred to make as less changes as possible relying on the fact that by default global structures are zeroed. So, acpi_capable will be default set to false if missing in the initialization.
> 
> Do you think we should explictly set it to false?

It should be explicitly false, yes. Relying on default for a positive information such as "This platform doesn't use ACPI" isn't good coding style. Neither case should be considered to be a semantic default.

File: include/sound/sof.h
Comment (plbossart) [inline include/sound/sof.h]: acpi_target_states_used ?

I feel acpi_capable is too broad.

**Review by plbossart — CHANGES_REQUESTED** [review]
yes @dbaluta this is what I had in mind, but I feel 'acpi_capable' is to vague/broad.

Question for @ranj063: would we limit legacy platforms to S0/S3?

**Comment by plbossart** [comment]
> > @paulstelian97 I preferred to make as less changes as possible relying on the fact that by default global structures are zeroed. So, acpi_capable will be default set to false if missing in the initialization.
> > Do you think we should explictly set it to false?
> 
> It should be explicitly false, yes. Relying on default for a positive information such as "This platform doesn't use ACPI" isn't good coding style. Neither case should be considered to be a semantic default.

I think we've crossed that bridge a long time ago and only set properties when they are not false.

**Comment by paulstelian97** [comment]
> > > @paulstelian97 I preferred to make as less changes as possible relying on the fact that by default global structures are zeroed. So, acpi_capable will be default set to false if missing in the initialization.
> > > Do you think we should explictly set it to false?
> > 
> > 
> > It should be explicitly false, yes. Relying on default for a positive information such as "This platform doesn't use ACPI" isn't good coding style. Neither case should be considered to be a semantic default.
> 
> I think we've crossed that bridge a long time ago and only set properties when they are not false.

Guess I shouldn't actually block on it, but I'll still say it's better to either explicitly initialize it or put a comment. It's clearer to do so.

File: include/sound/sof.h
Comment (dbaluta) [reply include/sound/sof.h]: or ```use_acpi_target_states```?

File: sound/soc/sof/pm.c
Comment (ranj063) [inline sound/soc/sof/pm.c]: @dbaluta this else is no longer needed isnt it?

**Comment by ranj063** [comment]
> yes @dbaluta this is what I had in mind, but I feel 'acpi_capable' is to vague/broad.
> 
> Question for @ranj063: would we limit legacy platforms to S0/S3?

@plbossart yes thats right but that's OK right?

**Comment by plbossart** [comment]
> > yes @dbaluta this is what I had in mind, but I feel 'acpi_capable' is to vague/broad.
> > Question for @ranj063: would we limit legacy platforms to S0/S3?
> 
> @plbossart yes thats right but that's OK right?

I just wanted to double check, I have zero desire to enable S0ix on those devices.

File: sound/soc/sof/pm.c
Comment (dbaluta) [reply sound/soc/sof/pm.c]: @ranj063 yes, I've seen this. But at least in the first version of the patch I wanted to keep it in order to have fewer changes. 

Also, care to explain a little bit what is acpi_target_system_state and what is system_suspend_target? Also, one can notice that if CONFIG_ACPI is not defined acpi_target_system_state by default is ACPI_STATE_S3.

**Review by plbossart — APPROVED** [review]

**Review by paulstelian97 — APPROVED** [review]
Still don't like the lack of explicit false properties but I guess I'm alone in that. Other than that, looks good.

**Comment by plbossart** [comment]
@ranj063 I'll let you merge if you are ok with the changes.

**Review by ranj063 — APPROVED** [review]

**Pull request merged**
